src/storage/compression/roaring.cpp

```diff
@@ -0,0 +1,1436 @@
+		} else {
+			amount = collection.cardinality[array_idx++];
+		}
+		return ContainerMetadata(is_inverted, is_run, amount);
+	}
+
+public:
+	const ContainerMetadataCollection &collection;
+	idx_t array_idx = 0;
+	idx_t run_idx = 0;
+	idx_t idx = 0;
+};
+
+struct RunContainerRLEPair {
+	uint16_t start;
+	uint16_t length;
+};
+
+enum class ContainerType : uint8_t { RUN_CONTAINER, ARRAY_CONTAINER, BITSET_CONTAINER };
+
+struct ContainerCompressionState {
+public:
+	struct Result {
+	public:
+		static Result RunContainer(uint16_t runs, bool nulls) {
+			auto res = Result();
+			res.container_type = ContainerType::RUN_CONTAINER;
+			res.nulls = nulls;
+			res.count = runs;
+			return res;
+		}
+
+		static Result ArrayContainer(uint16_t array_size, bool nulls) {
+			auto res = Result();
+			res.container_type = ContainerType::ARRAY_CONTAINER;
+			res.nulls = nulls;
+			res.count = array_size;
+			return res;
+		}
+
+		static Result BitsetContainer(uint16_t container_size) {
+			auto res = Result();
+			res.container_type = ContainerType::BITSET_CONTAINER;
+			res.nulls = true;
+			res.count = container_size;
+			return res;
+		}
+
+	public:
+		ContainerMetadata GetMetadata() const {
+			return ContainerMetadata(nulls, container_type == ContainerType::RUN_CONTAINER,
+			                         container_type == ContainerType::BITSET_CONTAINER ? MAX_ARRAY_IDX + 1 : count);
+		}
+		idx_t GetByteSize() const {
+			idx_t res = 0;
+			switch (container_type) {
+			case ContainerType::BITSET_CONTAINER:
+				res += AlignValue<idx_t, 8>(count) / 8;
+				break;
+			case ContainerType::RUN_CONTAINER:
+				res += count * sizeof(RunContainerRLEPair);
+				break;
+			case ContainerType::ARRAY_CONTAINER:
+				res += count * sizeof(uint16_t);
+				break;
+			}
+			return res;
+		}
+
+	public:
+		ContainerType container_type;
+		//! Whether nulls are being encoded or non-nulls
+		bool nulls;
+		//! The amount (meaning depends on container_type)
+		uint16_t count;
+
+	private:
+		Result() {
+		}
+	};
+
+public:
+	ContainerCompressionState() {
+		Reset();
+	}
+
+public:
+	void Append(bool null, uint16_t amount = 1) {
+		if (uncompressed) {
+			if (null) {
+				ValidityMask mask(uncompressed, ROARING_CONTAINER_SIZE);
+				SetInvalidRange(mask, count, count + amount);
+			}
+			count += amount;
+			return;
+		}
+
+		// Adjust the runs
+		auto &current_run_idx = run_idx[null];
+		auto &last_run_idx = run_idx[last_is_null];
+
+		if (count && null != last_is_null && last_run_idx < MAX_RUN_IDX) {
+			auto &last_run = runs[last_is_null][last_run_idx];
+			// End the last run
+			last_run.length = (count - last_run.start) - 1;
+			last_run_idx++;
+		}
+		if (!count || (null != last_is_null && current_run_idx < MAX_RUN_IDX)) {
+			auto &current_run = runs[null][current_run_idx];
+			// Initialize the new run
+			current_run.start = count;
+		}
+
+		// Add to the array
+		auto &current_array_idx = array_idx[null];
+		if (current_array_idx < MAX_ARRAY_IDX) {
```

Comment:
Should save roughly 31.25%, in some cases we might be worse off because of the bitpacking group size (32)
Meaning all our arrays/runs get aligned (ceiling) to multiples of 32

Reply:
For runs we don't really get a benefit though, as the biggest multiple that fits there is 64 (`93.09090` before floor)

Reply:
Since we're dealing with such small amounts of values, maybe we can remove this requirement of the bitpacking group size and use a less efficient procedure that doesn't assume multiples of 32

Or maybe we want to rethink the layout of the data, grouping multiple (all ?) vectors together to counteract the waste

Reply:
Just to document the new plan to compress arrays/runs:

For every range of 256 values in the container, which has as a size of 2048, we're adding a byte of overhead, for a total of 8. This applies to `array` and `run`.

For `array` this reduces the individual element cost from 16 bits (or 11 if we went with bitpacking) to 8 bits.
For `run` this reduces the individual element cost from 32 bits (or 22 if we went with bitpacking) to 16 bits.

### Compressed Array Containers

`array` already stored points inside the container.
Now an individual element in the array stores a point within the 256 value range that it falls into.
The extra byte for the 256 value range stores the count of points inside the range.

for example:
Uncompressed: 845
Compressed:
- encode the point:
(845 % 256) (`77`), counts[(845 / 256) (`3`)]++

### Compressed Run Containers

`run` currently stores (start, length-1) pairs for each run, where `start` is already a point in container.
To make use of the same encoding scheme we've also changed the length into a point in the container.

for example:
Uncompressed (start: 378, length: (200-1))
Compressed:
- convert `length` into an `end` point: (start: 378, end: 578)
- encode the points:
`start`: (378 % 256) (`122`), counts[(378 / 256) (`1`)]++
`end`: (578 % 256) (`66`), counts[(578 / 256) (`2`)]++